src/libexpr/primops/regex.cc

```diff
@@ -0,0 +1,360 @@
+        }
+    }
+}
+
+void prim_match(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
+
+    try {
+
+        auto& regex = state.regexCache->get(re);
+
+        PathSet context;
+        const auto str = state.forceString(*args[1], context, pos);
+
+        PCRE::MatchData match(regex);
+        int rc = regex.match(str, match, 0, PCRE2_ANCHORED | PCRE2_ENDANCHORED);
+        if (rc == PCRE2_ERROR_NOMATCH) {
+            v.mkNull();
+            return;
+        }
+        extract_matches(state, match, v);
+
+    } catch (RegexError & e) {
+        state.debugThrowLastTrace(EvalError({
+            .msg = hintfmt("error while evaluating regex '%s': ", re, e.what()),
+            .errPos = state.positions[pos]
+        }));
+    }
+}
+
+static RegisterPrimOp primop_match({
+    .name = "__match",
+    .args = {"regex", "str"},
+    .doc = R"s(
+      Returns a list if the [extended POSIX regular
+      expression](http://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap09.html#tag_09_04)
+      *regex* matches *str* precisely, otherwise returns `null`. Each item
+      in the list is a regex group.
+
+      ```nix
+      builtins.match "ab" "abc"
+      ```
+
+      Evaluates to `null`.
+
+      ```nix
+      builtins.match "abc" "abc"
+      ```
+
+      Evaluates to `[ ]`.
+
+      ```nix
+      builtins.match "a(b)(c)" "abc"
+      ```
+
+      Evaluates to `[ "b" "c" ]`.
+
+      ```nix
+      builtins.match "[[:space:]]+([[:upper:]]+)[[:space:]]+" "  FOO   "
+      ```
+
+      Evaluates to `[ "FOO" ]`.
+    )s",
+    .fun = prim_match,
+});
+
+/* Split a string with a regular expression, and return a list of the
+   non-matching parts interleaved by the lists of the matching groups. */
+void prim_split(EvalState & state, const PosIdx pos, Value * * args, Value & v)
+{
+    auto re = state.forceStringNoCtx(*args[0], pos);
+
+    try {
+        auto& regex = state.regexCache->get(re);
+
+        // We're going to use this regex many times, JIT it.
+        regex.compile();
+
+        PathSet context;
+        const auto str = state.forceString(*args[1], context, pos);
+
+
+        PCRE::MatchData match(regex);
+        int rc = regex.match(str, match);
+
+        if (rc == PCRE2_ERROR_NOMATCH) {
+            state.mkList(v, 1);
+            v.listElems()[0] = args[1];
+            return;
+        }
+        ValueVector result;
+        result.reserve(3);
+        size_t lastmatchend = 0;
+        size_t newmatchstart = 0;
+        while (rc != PCRE2_ERROR_NOMATCH && newmatchstart <= str.length()) {
+            // Add a string for non-matched characters.
+            Value * prefix = state.allocValue();
+            prefix->mkString(str.substr(lastmatchend, match.startPos() - lastmatchend));
+            result.push_back(prefix);
+
+            // Add a list for matched substrings.
+            auto elem = state.allocValue();
```

Comment:
the code is a bit inconsistent in choosing the type for `allocValue` calls. It's either `Value*` or `auto` and it would be nice to choose one variant and stick to it.